orca/service.go

```diff
@@ -81,32 +81,43 @@ type ServiceOptions struct {
 	allowAnyMinReportingInterval bool
 }
 
+// A ServerMetricsProvider provides ServerMetrics upon request.
+type ServerMetricsProvider interface {
+	// ServerMetrics returns the current set of server metrics.  It should
+	// return a read-only, immutable copy of the data that is active at the
+	// time of the call.
+	ServerMetrics() *ServerMetrics
+}
+
 // NewService creates a new ORCA service implementation configured using the
 // provided options.
 func NewService(opts ServiceOptions) (*Service, error) {
 	// The default minimum supported reporting interval value can be overridden
 	// for testing purposes through the orca internal package.
+	if opts.ServerMetricsProvider == nil {
+		return nil, fmt.Errorf("ServerMetricsProvider not specified")
+	}
 	if !opts.allowAnyMinReportingInterval {
 		if opts.MinReportingInterval < 0 || opts.MinReportingInterval < minReportingInterval {
 			opts.MinReportingInterval = minReportingInterval
 		}
 	}
 	service := &Service{
 		minReportingInterval: opts.MinReportingInterval,
-		utilization:          make(map[string]float64),
+		smProvider:           opts.ServerMetricsProvider,
 	}
 	return service, nil
 }
 
 // Register creates a new ORCA service implementation configured using the
 // provided options and registers the same on the provided service registrar.
-func Register(s *grpc.Server, opts ServiceOptions) (*Service, error) {
+func Register(s *grpc.Server, opts ServiceOptions) error {
```

Comment:
I see that this is existing code, but should/could we accept a `grpc.ServiceRegistrar` instead? If not, maybe we should fix the docstring.

Reply:
We needed the ORCA protos to be regenerated for this:

https://github.com/cncf/xds/issues/41

Updated the comment.